File: discipline_client_ts/rules copy/time_range_rule.ts
import { Branded, TimeRange, Countdown } from "../x.ts";

const BRAND = Symbol();

export type TimeRangeRule = Branded<typeof BRAND, {
  condition: TimeRange,
  lifetime: Countdown,
}>;

export const construct = (
  condition: TimeRange,
  lifetime: Countdown,
): TimeRangeRule => {
  return Branded(BRAND, {
    condition,
    lifetime,
  });
};

export const create = (
  timeRange: TimeRange,
  lifetime: Countdown,
): TimeRangeRule => {
  return construct(timeRange, lifetime);
};

export const getCondition = (it: TimeRangeRule): TimeRange => {
  return it.condition;
};

export const getLifetime = (it: TimeRangeRule): Countdown => {
  return it.lifetime;
};

export const TimeRangeRule = {
  construct,
  create,
  getCondition,
  getLifetime,
};

File: discipline4/app/src/main/java/com/example/app/rules/TimeRangeRule.kt
package com.example.app

import com.example.app.TimeRange
import com.example.app.Countdown
import androidx.room.Entity

/**
 * A rule that applies during a specific time range with a lifetime countdown
 */
@Entity
public data class TimeRangeRule private constructor(
  val condition: TimeRange,
  val lifetime: Countdown
) {
  companion object {
    fun create(timeRange: TimeRange, lifetime: Countdown): TimeRangeRule {
      return TimeRangeRule(timeRange, lifetime)
    }
  }
  
  fun getCondition(): TimeRange {
    return condition
  }

  fun getLifetime(): Countdown {
    return lifetime
  }
  
  /**
   * Checks if this rule is active at the given time
   */
  fun isActiveAt(instant: Instant, time: Time): Boolean {
    return condition.contains(time) && !lifetime.isFinished(instant)
  }
  
  /**
   * Checks if this rule has expired
   */
  fun isExpired(now: Instant): Boolean {
    return lifetime.isFinished(now)
  }
  
  /**
   * Gets the remaining time before this rule expires
   */
  fun getRemainingLifetime(now: Instant): Duration {
    return lifetime.getRemainingTimeOrZero(now)
  }
  
  override fun toString(): String {
    return "TimeRangeRule(condition=$condition, lifetime=${lifetime.getTotalDuration()})"
  }
}